pkg/volume/csi/csi_client.go

```diff
@@ -529,19 +534,33 @@ func asSingleNodeMultiWriterCapableCSIAccessModeV1(am api.PersistentVolumeAccess
 	return csipbv1.VolumeCapability_AccessMode_UNKNOWN
 }
 
-func newGrpcConn(addr csiAddr, metricsManager *MetricsManager) (*grpc.ClientConn, error) {
+func newGrpcConn(addr csiAddr, metricsManager *MetricsManager, tp trace.TracerProvider) (*grpc.ClientConn, error) {
 	network := "unix"
 	klog.V(4).InfoS(log("creating new gRPC connection"), "protocol", network, "endpoint", addr)
 
-	return grpc.Dial(
-		string(addr),
+	var dialOpts []grpc.DialOption
+	dialOpts = append(dialOpts,
 		grpc.WithAuthority("localhost"),
 		grpc.WithTransportCredentials(insecure.NewCredentials()),
 		grpc.WithContextDialer(func(ctx context.Context, target string) (net.Conn, error) {
 			return (&net.Dialer{}).DialContext(ctx, network, target)
 		}),
 		grpc.WithChainUnaryInterceptor(metricsManager.RecordMetricsInterceptor),
 	)
+
+	if tp != nil {
+		tracingOpts := []otelgrpc.Option{
+			otelgrpc.WithPropagators(tracing.Propagators()),
+			otelgrpc.WithTracerProvider(tp),
+		}
+		// Even if there is no TracerProvider, the otelgrpc still handles context propagation.
+		// See https://github.com/open-telemetry/opentelemetry-go/tree/main/example/passthrough
+		dialOpts = append(dialOpts,
```

Comment:
Should this be feature-gated?

Reply:
how about using the feature-gate `KubeletTracing`?

Reply:
That sounds good.

Reply:
Done